### ISSUE TITLE
Display track with multiple genres in each genre individually

### ISSUE THREAD

**Issue by Nozzie** [gerbera/gerbera#2405]
Is it possible to get Gerbera to display a track with multiple genres with the track into each separate genre?
For example, I have a track with three different genres. Now Gerbera concatenates the genres into one line.
I have `<upnp multi-value="yes"></upnp>` in the server part of the config.xml. I was expecting this to accomplish what I want.
See the images below for what happens now.

I would like to have Gerbera show three separate genres that all contain the track.

mp3tag display
![image](https://user-images.githubusercontent.com/5221405/152644062-5238f03b-76f4-44dd-a043-1dd8a8a1db5d.png)

Gerbera view:
![image](https://user-images.githubusercontent.com/5221405/152644024-df76ee46-746b-4616-88b3-f7169adb5083.png)

If I misconfigured something, please let me know.
Thanks

Edit:
I forgot to mention, I'm running in a docker container with `gerbera/gerbera:master`
version: v1.9.2-308-g6537a282

**Comment by Nozzie** [comment]
I did some testing, and I figured out the following.

With `<upnp multi-value="yes"></upnp>` in combination with `<virtual-layout type="builtin" audio-layout="Default">` both multiple artists and multiple genres do not end up in separate 'folders'.
When I set `<virtual-layout type="js" audio-layout="Default">`, multiple artist tags do get placed in separate 'folders', but genres are still concatenated. I added a second artist tag (test) to my test file.

![image](https://user-images.githubusercontent.com/5221405/152828447-667d1c8a-74d1-4051-b2b4-b47902229432.png)

I added the following code to import.js
```
for (var metaProp in obj.meta) {
    print('Property --->' + metaProp);
    print('Value ------>' + obj.meta[metaProp]);
    print('======================================');
}
```
That results in:
```
2022-02-07 16:11:07   info: Property --->dc:date
2022-02-07 16:11:07   info: Value ------>2002-01-01
2022-02-07 16:11:07   info: ======================================
2022-02-07 16:11:07   info: Property --->dc:title
2022-02-07 16:11:07   info: Value ------>I’m Not in Love
2022-02-07 16:11:07   info: ======================================
2022-02-07 16:11:07   info: Property --->upnp:album
2022-02-07 16:11:07   info: Value ------>10cc: The Singles (1975-1992)
2022-02-07 16:11:07   info: ======================================
2022-02-07 16:11:07   info: Property --->upnp:artist
2022-02-07 16:11:07   info: Value ------>10cc; test
2022-02-07 16:11:07   info: ======================================
2022-02-07 16:11:07   info: Property --->upnp:artist@role[AlbumArtist]
2022-02-07 16:11:07   info: Value ------>10cc
2022-02-07 16:11:07   info: ======================================
2022-02-07 16:11:07   info: Property --->upnp:date
2022-02-07 16:11:07   info: Value ------>2002-01-01
2022-02-07 16:11:07   info: ======================================
2022-02-07 16:11:07   info: Property --->upnp:episodeSeason
2022-02-07 16:11:07   info: Value ------>1
2022-02-07 16:11:07   info: ======================================
2022-02-07 16:11:07   info: Property --->upnp:genre
2022-02-07 16:11:07   info: Value ------>Ambient Pop Art Pop Pop Rock
2022-02-07 16:11:07   info: ======================================
2022-02-07 16:11:07   info: Property --->upnp:originalTrackNumber
2022-02-07 16:11:07   info: Value ------>2
2022-02-07 16:11:07   info: ======================================
```

Is the multi-value feature available for genres?

**Comment by KarlStraussberger** [comment]
If you run your script using `metaData` instead of `meta` you will get an array of values instead of a concatenated string.

To achieve your goal you need to loop over `metaData[M_GENRE]` 

```
    for (var oneGenre in metaData[M_GENRE]) {
            chain.genre.title = oneGenre ;
            chain.genre.metaData[M_GENRE] = [ oneGenre ];
            container = addContainerTree([chain.audio, chain.allGenres, chain.genre]);
            addCdsObject(obj, container);
    }
```

**Comment by Nozzie** [comment]
Thanks, I haven not got it to work yet, but at least I know it should be possible :)

Can you point me to where I should put that code block?

I tried putting it in import.js, but that gives `error: Failed to execute script: ReferenceError: identifier 'metaData' undefined`
If I change `metadata[M_GENRE]` to` obj.metaData[M_GENRE]` I get `error: Failed to execute script: ReferenceError: identifier 'chain' undefined`

So, basically I don't know where to put the code. I also tried to put it in the `addAudio` function in common.js without success.

**Comment by Nozzie** [comment]
For me, both obj and orig.obj.metaData[M_GENRE] already contain a concatenated string of multiple genre fields.
```
2022-02-12 09:45:51   info: Value obj.metaData[M_GENRE] --> Ambient Pop Art Pop Pop Rock
2022-02-12 09:45:51   info: Value orig.metaData[M_GENRE] --> Ambient Pop Art Pop Pop Rock
```

**Comment by KarlStraussberger** [comment]
The taglib code still uses the property `genre()`. Need to change that to list handling.

**Comment by Nozzie** [comment]
In the meanwhile, I found a workaround for this issue. 

**Prerequisites:**
Tracks with multiple genre fields have to be retagged so they contain a single genre field, where multiple values are delimited with a delimiter of choice (I used ";").

Add the following code to import.js, just below `case "audio":`
```
            const g_array = obj.metaData[M_GENRE][0].split(";");
            for (var g in g_array) {
                obj.metaData[M_GENRE][g] = g_array[g].trim();
            }
```
Change function addAudio to the following:
```
// doc-add-audio-begin
function addAudio(obj) {
    // Note the difference between obj.title and obj.metaData[M_TITLE] -
    // while object.title will originally be set to the file name,
    // obj.metaData[M_TITLE] will contain the parsed title - in this
    // particular example the ID3 title of an MP3.
    var title = obj.title;

    // First we will gather all the metadata that is provided by our
    // object, of course it is possible that some fields are empty -
    // we will have to check that to make sure that we handle this
    // case correctly.
    if (obj.metaData[M_TITLE] && obj.metaData[M_TITLE][0]) {
        title = obj.metaData[M_TITLE][0];
    }

    var desc = '';
    var artist = [ 'Unknown' ];
    var artist_full = null;
    if (obj.metaData[M_ARTIST] && obj.metaData[M_ARTIST][0]) {
        artist = obj.metaData[M_ARTIST];
        artist_full = artist.join(' / ');
        desc = artist_full;
    }

    var album = 'Unknown';
    var album_full = null;
    if (obj.metaData[M_ALBUM] && obj.metaData[M_ALBUM][0]) {
        album = obj.metaData[M_ALBUM][0];
        desc = desc + ', ' + album;
        album_full = album;
    }

    if (desc) {
        desc = desc + ', ';
    }
    desc = desc + title;

    var date = 'Unknown';
    if (obj.metaData[M_DATE] && obj.metaData[M_DATE][0]) {
        date = getYear(obj.metaData[M_DATE][0]);
        obj.metaData[M_UPNP_DATE] = [ date ];
        desc = desc + ', ' + date;
    }

    var genre = [ 'Unknown' ];
    if (obj.metaData[M_GENRE] && obj.metaData[M_GENRE][0]) {
        genre = mapGenre(obj.metaData[M_GENRE]);
        desc = desc + ', ' + genre.join(' / ');
    }

    if (!obj.metaData[M_DESCRIPTION] || !obj.metaData[M_DESCRIPTION][0]) {
        obj.description = desc;
    }

    var composer = 'None';
    if (obj.metaData[M_COMPOSER] && obj.metaData[M_COMPOSER][0]) {
        composer = obj.metaData[M_COMPOSER].join(' / ');
    }

/*
    var conductor = 'None';
    if (obj.metaData[M_CONDUCTOR] && obj.metaData[M_CONDUCTOR][0]) {
        conductor = obj.metaData[M_CONDUCTOR].join(' / ');
    }

    var orchestra = 'None';
    if (obj.metaData[M_ORCHESTRA] && obj.metaData[M_ORCHESTRA][0]) {
        orchestra = obj.metaData[M_ORCHESTRA].join(' / ');
    }
*/
    // uncomment this if you want to have track numbers in front of the title
    // in album view
/*
    var track = '';
    if (obj.metaData[M_TRACKNUMBER] && obj.metaData[M_TRACKNUMBER][0]) {
        track = obj.metaData[M_TRACKNUMBER][0];
        if (track.length == 1) {
            track = '0' + track;
        }
        track = track + ' ';
    }
*/
    // comment the following line out if you uncomment the stuff above  :)
    var track = '';

// uncomment this if you want to have channel numbers in front of the title
/*
    var channels = obj.res[R_NRAUDIOCHANNELS];
    if (channels) {
        if (channels === "1") {
            track = track + '[MONO]';
        } else if (channels === "2") {
            track = track + '[STEREO]';
        } else {
            track = track + '[MULTI]';
        }
    }
*/

    // We finally gathered all data that we need, so let's create a
    // nice layout for our audio files.

    obj.title = title;

    // The UPnP class argument to addCdsObject() is optional, if it is
    // not supplied the default UPnP class will be used. However, it
    // is suggested to correctly set UPnP classes of containers and
    // objects - this information may be used by some renderers to
    // identify the type of the container and present the content in a
    // different manner.

    // Remember, the server will sort all items by ID3 track if the
    // container class is set to UPNP_CLASS_CONTAINER_MUSIC_ALBUM.

    const chain = {
        audio: { title: 'Audio', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER, metaData: [] },
        allAudio: { title: 'All Audio', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER },
        allArtists: { title: 'Artists', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER },
        allGenres: { title: 'Genres', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER },
        allAlbums: { title: 'Albums', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER },
        allYears: { title: 'Year', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER },
        allComposers: { title: 'Composers', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER },
        allSongs: { title: 'All Songs', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER },
        allFull: { title: 'All - full name', objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER },
        artist: { searchable: false, title: artist[0], objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER_MUSIC_ARTIST, metaData: [], res: obj.res, aux: obj.aux, refID: obj.id },
        album: { searchable: false, title: album, objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER_MUSIC_ALBUM, metaData: [], res: obj.res, aux: obj.aux, refID: obj.id },
        genre: { title: genre, objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER_MUSIC_GENRE, metaData: [], res: obj.res, aux: obj.aux, refID: obj.id },
        year: { title: date, objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER, metaData: [], res: obj.res, aux: obj.aux, refID: obj.id },
        composer: { title: composer, objectType: OBJECT_TYPE_CONTAINER, upnpclass: UPNP_CLASS_CONTAINER_MUSIC_COMPOSER, metaData: [], res: obj.res, aux: obj.aux, refID: obj.id },
    };

    chain.audio.metaData[M_CONTENT_CLASS] = [ UPNP_CLASS_AUDIO_ITEM ];
    chain.album.metaData[M_ARTIST] = artist;
    chain.album.metaData[M_ALBUMARTIST] = artist;
    chain.album.metaData[M_GENRE] = [ genre ];
    chain.album.metaData[M_DATE] = obj.metaData[M_DATE];
    chain.album.metaData[M_UPNP_DATE] = obj.metaData[M_UPNP_DATE];
    chain.album.metaData[M_ALBUM] = [ album ];
    chain.artist.metaData[M_ARTIST] = artist;
    chain.artist.metaData[M_ALBUMARTIST] = artist;
    chain.genre.metaData[M_GENRE] = genre;
    chain.year.metaData[M_DATE] = [ date ];
    chain.year.metaData[M_UPNP_DATE] = [ date ];
    chain.composer.metaData[M_COMPOSER] = [ composer ];

    var container = addContainerTree([chain.audio, chain.allAudio]);
    addCdsObject(obj, container);

    container = addContainerTree([chain.audio, chain.allArtists, chain.artist, chain.allSongs]);
    addCdsObject(obj, container);

    var temp = '';
    if (artist_full) {
        temp = artist_full;
    }

    if (album_full) {
        temp = temp + ' - ' + album_full + ' - ';
    } else {
        temp = temp + ' - ';
    }

    obj.title = temp + title;
    container = addContainerTree([chain.audio, chain.allFull]);
    addCdsObject(obj, container);

    const artCnt = artist.length;
    var i;
    for (i = 0; i < artCnt; i++) {
        chain.artist.title = artist[i];
        container = addContainerTree([chain.audio, chain.allArtists, chain.artist, chain.allFull]);
        addCdsObject(obj, container);
    }

    obj.title = track + title;
    for (i = 0; i < artCnt; i++) {
        chain.artist.title = artist[i];
        chain.artist.searchable = true;
        container = addContainerTree([chain.audio, chain.allArtists, chain.artist, chain.album]);
        addCdsObject(obj, container);
    }

    chain.album.searchable = true;
    container = addContainerTree([chain.audio, chain.allAlbums, chain.album]);
    obj.title = track + title;
    addCdsObject(obj, container);

    chain.genre.searchable = true;
//    var allGenre = getCombinations(genre);
    var allGenre = genre;
    const genCnt = allGenre.length;
    var i;
    for (i = 0; i < genCnt; i++) {
        chain.genre.title = allGenre[i];
        container = addContainerTree([chain.audio, chain.allGenres, chain.genre]);
        addCdsObject(obj, container);
    }

    container = addContainerTree([chain.audio, chain.allYears, chain.year]);
    addCdsObject(obj, container);

    container = addContainerTree([chain.audio, chain.allComposers, chain.composer]);
    addCdsObject(obj, container);
}
// doc-add-audio-end
```
Result:
![image](https://user-images.githubusercontent.com/5221405/153713293-d18e37f0-f1c8-4fa7-b0de-0422a5b2e7ea.png)

If you also want the unique combinations of the genres, comment `var allGenre = genre;` and uncomment `//    var allGenre = getCombinations(genre);`.
Also add the following function to common.js
```
// Create all possible combinations of array values
function getCombinations(items) {
  var result = [];
  var f = function(prefix, items) {
    for (var i = 0; i < items.length; i++) {
      if (prefix == '') {
        result.push(prefix.trim() + items[i].trim());
      } else {
          result.push(prefix.trim() + " " + items[i].trim());
      }
      f(prefix +  " " + items[i], items.slice(i + 1));
    }
  }
  f('', items.sort());
  return result;
}
```
 
Result:
![image](https://user-images.githubusercontent.com/5221405/153713263-946e3b7b-998a-47cf-be49-0a4152e8a479.png)

 Enjoy!

**Comment by KarlStraussberger** [comment]
I created a PR #2428 should solve your problem and read the genre values correctly.

**Comment by Nozzie** [comment]
Thanks, it works perfectly.

For people that want to also create all possible combinations of Genres, below is the import.js to accomplish that.
It handles both single Genre tags with delimited values, and multiple Genre tags.
It is no longer necessary to modify common.js 

```
// main script part

/*
 Create all possible combinations of Genre values
 You probably should not use this function if you have more than three Genres per track
 The number of possible combinations increases exponetially
*/
function getCombinations(items) {

/*
   Delimiter used to combine genres in the virtual container
   This will be used in the vidible result in Gerbera
   Genre = 'Rock/Hard Rock/Metal'
*/
  var displayDelimiter = "/";

/*
   Delimiter used in single field Genre tags to hold multiple values
   This is what is read from the audio file
   Genre = 'Rock;Hard Rock;Metal'
*/
  var multiDelimiter = ';';

  // Check if any of the tags has a delimiter, if so, split on delimiter and recreate items array
  for (var i = 0; i < items.length; i++) {
        if (items[i].indexOf(multiDelimiter, 1) > 1) {
            const spltArr = items[i].split(multiDelimiter);
            items[i] = spltArr[0];
            items =  items.concat(spltArr.slice(1));
        }
  }
  items.sort();

  var result = [];
  var f = function(prefix, items) {
    for (var i = 0; i < items.length; i++) {
        if (prefix == '') {
            result.push(items[i].trim());
        } else {
            if (prefix.charAt(0) == displayDelimiter) {
                result.push(prefix.replace(displayDelimiter, '') + displayDelimiter + items[i].trim());
            } else {
                result.push(prefix + displayDelimiter + items[i].trim());
            }
        }
        f(prefix +  displayDelimiter + items[i], items.slice(i + 1));
    }
  }
  f('', items);
  return result;
}

if (getPlaylistType(orig.mimetype) === '') {
    var arr = orig.mimetype.split('/');
    var mime = arr[0];

        // All virtual objects are references to objects in the
    // PC-Directory, so make sure to correctly set the reference ID!
    var obj = orig;
    obj.refID = orig.id;

    if (mime === 'video' && obj.onlineservice === ONLINE_SERVICE_APPLE_TRAILERS) {
        mime = 'trailer';
    } else if (orig.mimetype === 'application/ogg') {
        mime = (orig.theora === 1) ? 'video' : 'audio';
    }
    var audioLayout = config['/import/scripting/virtual-layout/attribute::audio-layout'];
    if (!audioLayout) {
        audioLayout = 'Default';
    }
    switch (mime) {
        case "audio":
            obj.metaData[M_GENRE] =  getCombinations(obj.metaData[M_GENRE]);

            switch (audioLayout) {
                case 'Structured':
                    addAudioStructured(obj);
                    break;
                default:
                    addAudio(obj);
                    break;
            }
            break;
        case "video":
            addVideo(obj);
            break;
        case "trailer":
            addTrailer(obj);
            break;
        case "image":
            addImage(obj);
            break;
        default:
            print("Unable to handle mime type " + orig.mimetype + " for " + obj.location);
            break;
    }
}

```

**Comment by whyman** [comment]
Thanks for testing @Nozzie!
